internal/pkg/manifest/union.go

```diff
@@ -0,0 +1,163 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package manifest
+
+import (
+	"gopkg.in/yaml.v3"
+)
+
+// Union is a type used for yaml keys that may be of type A or B.
+// Union will only ever hold one of the underlying types, never both.
+//
+// Use NewUnionA() and NewUnionB() to create a Union with an underlying
+// type already set. See Unmarshal() for details on how yaml is decoded
+// into Union.
+//
+// Union is exported to enable type embedding.
+type Union[A, B any] struct {
+	// isA is true if the underlying type of Union is A.
+	isA bool
```

Comment:
Something like `Short` vs `AdvancedConfig` or anything equivalent.  At any rate it is `manifest.Union` so it is possible that this type is not an absolutely general `Union`.

Reply:
discussed in https://github.com/aws/copilot-cli/pull/4100#discussion_r1001974462!